From 7f128168879749554062fbdcc835cdeac4c76032 Mon Sep 17 00:00:00 2001
From: Jyotiraditya Panda <jyotiraditya@aospa.co>
Date: Wed, 7 Aug 2024 21:35:48 +0530
Subject: [PATCH 06/16] PropImitationHooks: Allow user to disable prop
 imitation

 - Prevents spoofing props, bypassing task perms and blocking key attestation for GMS.

[adithya2306: add a prop for disabling only key attestation too]

Change-Id: I529e91e8d11a27d064c3ef85a8cf5e73e1857e3b
Signed-off-by: Jyotiraditya Panda <jyotiraditya@aospa.co>
Co-authored-by: Adithya R <gh0strider.2k18.reborn@gmail.com>
---
 .../internal/util/PropImitationHooks.java     | 26 ++++++++++++++++++-
 1 file changed, 25 insertions(+), 1 deletion(-)

diff --git a/core/java/com/android/internal/util/PropImitationHooks.java b/core/java/com/android/internal/util/PropImitationHooks.java
index da1e226a5ada..c8c13fe5ee4b 100644
--- a/core/java/com/android/internal/util/PropImitationHooks.java
+++ b/core/java/com/android/internal/util/PropImitationHooks.java
@@ -42,7 +42,13 @@ import java.util.Set;
 public class PropImitationHooks {
 
     private static final String TAG = "PropImitationHooks";
-    private static final boolean DEBUG = false;
+    private static final boolean DEBUG = Log.isLoggable(TAG, Log.DEBUG);
+
+    private static final Boolean sDisableGmsProps = SystemProperties.getBoolean(
+            "persist.sys.pihooks.disable.gms_props", false);
+
+    private static final Boolean sDisableKeyAttestationBlock = SystemProperties.getBoolean(
+            "persist.sys.pihooks.disable.gms_key_attestation_block", false);
 
     private static final String PACKAGE_ARCORE = "com.google.ar.core";
     private static final String PACKAGE_FINSKY = "com.android.vending";
@@ -144,10 +150,19 @@ public class PropImitationHooks {
     }
 
     private static void setCertifiedPropsForGms() {
+        if (sDisableGmsProps) {
+            dlog("GMS prop imitation is disabled by user");
+            setSystemProperty(PROP_SECURITY_PATCH, Build.VERSION.SECURITY_PATCH);
+            setSystemProperty(PROP_FIRST_API_LEVEL,
+                    Integer.toString(Build.VERSION.DEVICE_INITIAL_SDK_INT));
+            return;
+        }
+
         if (sCertifiedProps.length == 0) {
             dlog("Certified props are not set");
             return;
         }
+
         final boolean was = isGmsAddAccountActivityOnTop();
         final TaskStackListener taskStackListener = new TaskStackListener() {
             @Override
@@ -210,6 +225,10 @@ public class PropImitationHooks {
     }
 
     public static boolean shouldBypassTaskPermission(Context context) {
+        if (sDisableGmsProps) {
+            return false;
+        }
+
         // GMS doesn't have MANAGE_ACTIVITY_TASKS permission
         final int callingUid = Binder.getCallingUid();
         final int gmsUid;
@@ -229,6 +248,11 @@ public class PropImitationHooks {
     }
 
     public static void onEngineGetCertificateChain() {
+        if (sDisableKeyAttestationBlock) {
+            dlog("Key attestation blocking is disabled by user");
+            return;
+        }
+
         // Check stack for SafetyNet or Play Integrity
         if (isCallerSafetyNet() || sIsFinsky) {
             dlog("Blocked key attestation sIsGms=" + sIsGms + " sIsFinsky=" + sIsFinsky);
-- 
2.46.2

